File: vstidrv3/winmmhelp.cpp
#include "winmmhelp.h"

namespace Boare{ namespace Cadencii{

    unsigned int winmmhelp::s_num_joydev = 0;
    bool *winmmhelp::s_joy_attatched;
    int *winmmhelp::s_button_num;
    bool winmmhelp::s_initialized = false;
    int *winmmhelp::s_joy_available;

    int winmmhelp::Init(){
        if( s_initialized ){
            Reset();
        }
        s_initialized = true;
        int num_joydev = joyGetNumDevs();
        if ( num_joydev <= 0 ) {
            num_joydev = 0;
            return num_joydev;
        }
        s_joy_attatched = new bool[num_joydev];
        s_button_num = new int[num_joydev];
        int count = 0;
        for( int k = 0; k < num_joydev; k++ ){
            JOYINFO ji;
            if( joyGetPos( k, &ji ) == JOYERR_NOERROR ){
                s_joy_attatched[k] = true;
                JOYCAPS jc;
                joyGetDevCaps( k, &jc, sizeof( JOYCAPS ) );
                s_button_num[k] = jc.wNumButtons;
                count++;
            }else{
                s_joy_attatched[k] = false;
                s_button_num[k] = 0;
            }
        }
        if( count > 0 ){
            s_joy_available = new int[count];
            int c = -1;
            for( int i = 0; i < num_joydev; i++ ){
                if( s_joy_attatched[i] ){
                    c++;
                    if( c >= count ){
                        break; //Ç±Ç±Ç…óàÇÈÇÃÇÕÉGÉâÅ[
                    }
                    s_joy_available[c] = i;
                }
            }
        }
        s_num_joydev = count;
        return s_num_joydev;
    };

    bool winmmhelp::IsJoyAttatched( int index ){
        if( !s_initialized ){
            Init();
        }
        if( s_num_joydev == 0 || index < 0 || (int)s_num_joydev <= index ){
            return false;
        }
        return s_joy_attatched[index];
    };
    
    bool winmmhelp::GetStatus( int index_, unsigned char *buttons, int len, int *pov ){
        const int _BTN[32] = { 0x0001,
                               0x0002,
                               0x0004,
                               0x0008,
                               0x00000010l,
                               0x00000020l,
                               0x00000040l,
                               0x00000080l,
                               0x00000100l,
                               0x00000200l,
                               0x00000400l,
                               0x00000800l,
                               0x00001000l,
                               0x00002000l,
                               0x00004000l,
                               0x00008000l,
                               0x00010000l,
                               0x00020000l,
                               0x00040000l,
                               0x00080000l,
                               0x00100000l,
                               0x00200000l,
                               0x00400000l,
                               0x00800000l,
                               0x01000000l,
                               0x02000000l,
                               0x04000000l,
                               0x08000000l,
                               0x10000000l,
                               0x20000000l,
                               0x40000000l,
                               0x80000000l };
        if( !s_initialized ){
            return false;
        }
        if( s_num_joydev == 0 || index_ < 0 || (int)s_num_joydev <= index_ ){
            return false;
        }
        int index = s_joy_available[index_];
        JOYINFOEX ji_ex;
        ji_ex.dwSize = sizeof( JOYINFOEX );
        ji_ex.dwFlags = JOY_RETURNPOV | JOY_RETURNBUTTONS;

        if( s_joy_attatched[index] ){
            joyGetPosEx( index, &ji_ex );
            *pov = ji_ex.dwPOV;
            if( *pov == 0xffff ){
                *pov = -1;
            }
            for( int i = 0; i < len && i < s_button_num[index]; i++ ){
                buttons[i] = (ji_ex.dwButtons & _BTN[i]) ? 0x80 : 0x00;
            }
            return true;
        }else{
            return false;
        }
    };

    int winmmhelp::GetNumButtons( int index ){
        if( !s_initialized ){
            Init();
        }
        if( s_num_joydev == 0 || index < 0 || (int)s_num_joydev <= index ){
            return 0;
        }
        return s_button_num[s_joy_available[index]];
    };

    void winmmhelp::Reset(){
        if( s_initialized ){
            delete [] s_button_num;
            delete [] s_joy_attatched;
            delete [] s_joy_available;
            s_initialized = false;
            s_num_joydev = 0;
        }
    };

    int winmmhelp::GetNumJoyDev(){
        if( !s_initialized ){
            return Init();
        }else{
            return s_num_joydev;
        }
    };

} }

/*int main(){
    Boare::Cadencii::winmmhelp::Init();
    std::cout << "main" << std::endl;
    unsigned char *status;
    int num_joy = Boare::Cadencii::winmmhelp::Init();
    int num_button;
    if( num_joy <= 0 ){
        std::cout << "num_joy=" << num_joy << std::endl;
        goto final;
    }
    std::cout << "num_joy=" << num_joy << std::endl;
    for( int k = 0; k < num_joy; k++ ){
        std::cout << "winmmhelp::IsJoy" << k << "Attatched=" << Boare::Cadencii::winmmhelp::IsJoyAttatched( k ) << std::endl;
    }
    num_button = Boare::Cadencii::winmmhelp::GetNumButtons( 0 );
    std::cout << "num_button=" << num_button << std::endl;
    if( num_button <= 0 ){
        goto final;
    }
    status = new unsigned char[num_button];
    int pov;
    while( true ){
        Boare::Cadencii::winmmhelp::GetStatus( 0, status, num_button, &pov );
        std::cout << "\r" << std::dec << std::setw( 5 ) << pov << std::hex;
        for( int i = 0; i < num_button; i++ ){
            int v = status[i];
            std::cout << " 0x" << std::setw( 2 ) << std::setfill( '0' ) << v;
        }
    }
final:
    if( status ){
        delete [] status;
    }
    int i;
    std::cin >> i;
    return 0;
}*/

File: Cadencii/winmmhelp.cs
/*
 * winmmhelp.cs
 * Copyright (c) 2009 kbinani
 *
 * This file is part of Boare.Cadencii.
 *
 * Boare.Cadencii is free software; you can redistribute it and/or
 * modify it under the terms of the GPLv3 License.
 *
 * Boare.Cadencii is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 */
using System;
using System.Runtime.InteropServices;

using bocoree;

namespace Boare.Cadencii {

    using boolean = Boolean;

    public unsafe static class winmmhelp {
        static uint s_num_joydev = 0;
        static boolean[] s_joy_attatched;
        static int[] s_button_num;
        static boolean s_initialized = false;
        static int[] s_joy_available;

        public static int JoyInit() {
            if ( s_initialized ) {
                JoyReset();
            }
            s_initialized = true;
            int num_joydev = (int)windows.joyGetNumDevs();
#if DEBUG
            bocoree.debug.push_log( "winmmhelp.JoyInit" );
            bocoree.debug.push_log( "    num_joydev=" + num_joydev );
#endif
            if ( num_joydev <= 0 ) {
                num_joydev = 0;
                return num_joydev;
            }
            s_joy_attatched = new boolean[num_joydev];
            s_button_num = new int[num_joydev];
            int count = 0;
            for ( int k = 0; k < num_joydev; k++ ) {
                JOYINFO ji = new JOYINFO();
                if ( windows.joyGetPos( (uint)k, ref ji ) == windows.JOYERR_NOERROR ) {
                    s_joy_attatched[k] = true;
                    JOYCAPSW jc = new JOYCAPSW();
                    windows.joyGetDevCapsW( (uint)k, ref jc, (uint)Marshal.SizeOf( jc ) );
                    s_button_num[k] = (int)jc.wNumButtons;
                    count++;
                } else {
                    s_joy_attatched[k] = false;
                    s_button_num[k] = 0;
                }
            }
            if ( count > 0 ) {
                s_joy_available = new int[count];
                int c = -1;
                for ( int i = 0; i < num_joydev; i++ ) {
                    if ( s_joy_attatched[i] ) {
                        c++;
                        if ( c >= count ) {
                            break; //ここに来るのはエラー
                        }
                        s_joy_available[c] = i;
                    }
                }
            }
            s_num_joydev = (uint)count;
            return (int)s_num_joydev;
        }

        public static boolean IsJoyAttatched( int index ) {
            if ( !s_initialized ) {
                JoyInit();
            }
            if ( s_num_joydev == 0 || index < 0 || (int)s_num_joydev <= index ) {
                return false;
            }
            return s_joy_attatched[index];
        }

        public static boolean JoyGetStatus( int index_, out byte[] buttons, out int pov ) {
            uint[] _BTN = new uint[32] { 0x0001,
                               0x0002,
                               0x0004,
                               0x0008,
                               0x00000010,
                               0x00000020,
                               0x00000040,
                               0x00000080,
                               0x00000100,
                               0x00000200,
                               0x00000400,
                               0x00000800,
                               0x00001000,
                               0x00002000,
                               0x00004000,
                               0x00008000,
                               0x00010000,
                               0x00020000,
                               0x00040000,
                               0x00080000,
                               0x00100000,
                               0x00200000,
                               0x00400000,
                               0x00800000,
                               0x01000000,
                               0x02000000,
                               0x04000000,
                               0x08000000,
                               0x10000000,
                               0x20000000,
                               0x40000000,
                               0x80000000 };
            if ( !s_initialized ) {
                pov = -1;
                buttons = new byte[0];
                return false;
            }
            if ( s_num_joydev == 0 || index_ < 0 || (int)s_num_joydev <= index_ ) {
                pov = -1;
                buttons = new byte[0];
                return false;
            }
            int len = s_button_num[index_];
            buttons = new byte[len];
            pov = -1;
            int index = s_joy_available[index_];
            JOYINFOEX ji_ex = new JOYINFOEX();
            ji_ex.dwSize = (ushort)Marshal.SizeOf( ji_ex );
            ji_ex.dwFlags = windows.JOY_RETURNPOV | windows.JOY_RETURNBUTTONS;

            if ( s_joy_attatched[index] ) {
                windows.joyGetPosEx( (uint)index, ref ji_ex );
                pov = (int)ji_ex.dwPOV;
                if ( pov == 0xffff ) {
                    pov = -1;
                }
                for ( int i = 0; i < len && i < s_button_num[index]; i++ ) {
                    buttons[i] = (((uint)ji_ex.dwButtons & _BTN[i]) != 0x0) ? (byte)0x80 : (byte)0x00;
                }
                return true;
            } else {
                return false;
            }
        }

        public static int JoyGetNumButtons( int index ) {
            if ( !s_initialized ) {
                JoyInit();
            }
            if ( s_num_joydev == 0 || index < 0 || (int)s_num_joydev <= index ) {
                return 0;
            }
            return s_button_num[s_joy_available[index]];
        }

        public static void JoyReset() {
            if ( s_initialized ) {
                s_initialized = false;
                s_num_joydev = 0;
            }
        }

        public static int JoyGetNumJoyDev() {
            if ( !s_initialized ) {
                return JoyInit();
            } else {
                return (int)s_num_joydev;
            }
        }
    }

}
